Graph theory in Nit

ablondin opened the pull request:
As mentioned in issue #1386, I would like to introduce a graph theory module in Nit.

@Morriar : any comment about the documentation style is welcome!

Task list:

- [x] Should be `AbstractDigraph[V]` instead of `AbstractDigraph[V: Object]`
- [x] No newline in `to_s`
- [x] Call `escape_to_dot` in `to_graphviz`
- [x] Use a more efficient data structure for the queue in `a_shortest_path`
- [x] `HashSet[V]` instead of `HashMap[V, Bool]`
- [x] Change `loop-end` by `while-do-end` in `a_shortest_path` and `distance` methods
- [x] `index_to_vertex: HashMap[Int, V]` should be `index_to_vertex: Array[V]`
- [x] Remove useless `QueuePair` class
- [x] Rename `to_graphviz` by `to_dot`
- [x] One-line short functions
- [x] Change `vertices` method's signature (should return `Collection[V]`)
- [x] `num_vertices` and `num_arcs` should be abstract methods
- [x] Divide the services between *read-only* and *read/write*
- [x] Split modules between abstract and concrete classes
- [x] Document all methods
- [x] Split methods groups more carefully
- [x] Make `V` nullable everywhere
- [x] Remove `digraph_test.nit` file
- [x] Adds an `intern_get_arc` (cleaner + factor code)
- [x] Behavior of `add_arc` (in particular with label) should be made clear
- [x] Make `arc_value` function more general (setter + getter)
- [x] Reading arcs, incoming/outgoing arcs safer (readable only)
- [x] Adds `is_empty` and `iterator` methods
- [x] Move ``arcs``, ``incoming_arcs`` and ``outgoing_arcs`` in the abstract class
- [x] Rename classes (``AbstractDigraph -> Digraph`` and ``HashMapDigraph -> HashDigraph``)
- [x] Document the modules with overview + examples
- [x] Move Tarjan's algorithm to a private class

privat commented on lib/graphs/digraph.nit:
```diff
@@ -0,0 +1,606 @@
+end
+
+# Interface for digraphs
+abstract class AbstractDigraph[V: Object]
```
Why a restriction on Object instead of nullable Object?

privat commented on lib/graphs/digraph.nit:
```diff
@@ -0,0 +1,606 @@
+		var t = "arcs"
+		if num_vertices <= 1 then s = "vertex"
+		if num_arcs <= 1 then t = "arc"
+		s = "Digraph of {num_vertices} {s} and {num_arcs} {t}\n"
```
I am not a fan of having newlines in `to_s` because it may produce bad results when combined by clients that call `to_s` in an abstract way. It is just a gut feeling so I wont complain more.

privat commented on lib/graphs/digraph.nit:
```diff
@@ -0,0 +1,606 @@
+		var s = "digraph \{\n"
+		# Writing the vertices
+		for u in vertices do
+			s += "  \"{u}\" [label=\"{u}\"];\n"
```
Use `escape_to_dot` (in standard) to automatically escape things and produce a valid graphviz file.

~~~nit
s += "  \"{u.escape_to_dot}\" [label=\"{u.escape_to_dot}\"];\n"
~~~

privat commented on lib/graphs/digraph.nit:
```diff
@@ -0,0 +1,606 @@
+	# If no path exists between `u` and `v`, it returns `null`.
+	fun a_shortest_path(u, v: V): nullable List[V]
+	do
+		var queue = (new Array[V]).as_fifo
```
from the doc:

~~~
# Note that some `Sequence`, like `Array`, have an inefficient `shift` implementation,
# thus are not the best candidates to serve as a FIFO.
~~~

privat commented on lib/graphs/digraph.nit:
```diff
@@ -0,0 +1,606 @@
+	do
+		var queue = (new Array[V]).as_fifo
+		var pred = new HashMap[V, nullable V]
+		var visited = new HashMap[V, Bool]
```
HashSet[V] instead of  HashMap[V, Bool]?

privat commented on lib/graphs/digraph.nit:
```diff
@@ -0,0 +1,606 @@
+		queue.add(u)
+		loop
+			w = queue.take
+			if not visited.get_or_default(w, false) then
```
~~~nit
if visited.has(w) then continue
~~~

privat commented on lib/graphs/digraph.nit:
```diff
@@ -0,0 +1,606 @@
+		var w: nullable V
+		pred[u] = null
+		queue.add(u)
+		loop
```
~~~nit
while queue.not_empty do
~~~

privat commented on lib/graphs/digraph.nit:
```diff
@@ -0,0 +1,606 @@
+
+	# Attributes
+	private var pred_map = new HashMap[V, Array[V]]
+	private var succ_map = new HashMap[V, Array[V]]
```
Personally, I would have reified the node with a private structure to factorize things.

~~~nit
private var nodes = new HashMap[V, HashMapNode[[V]]
#...
private class HashMapNode [V]
   var preds = new Array[V]
   var succs = new Array[V]
end
~~~

This add an indirection level but has the advantage to merge the two hashmaps.
Another benefit is that you have a private class for nodes that you can use to store other data if you want.

So there is pros and cons. I do not know it this worth the change, or even some objective measure of the cost of each one.

privat commented on lib/graphs/digraph.nit:
```diff
@@ -0,0 +1,606 @@
+
+	redef fun vertices: RemovableCollection[V]
+	do
+		return succ_map.keys
```
because you return a Removable collection, people may invalidate the data structure because modification on `succ_map.keys` will not be applied on `pred_map` or `num_vertices`.

privat replied:
A solution could be to return a Collection (thus hide mutation methods)

ablondin replied:
Do I have to cast it or just change the method's signature?

privat replied:
Changing the method signature should be enough.

privat commented on lib/graphs/digraph.nit:
```diff
@@ -0,0 +1,606 @@
+
+	# Attributes
+	private var vertex_to_index = new HashMap[V, Int]
+	private var index_to_vertex = new HashMap[Int, V]
```
since the key ranges in `[0..num_vertices[` , I suggest an array instead (`Array[V]`)

ablondin replied:
@privat: This was my first idea, but I think it would be inefficient, since we can remove keys from the middle of the array, so that shifting all values could be expensive. Don't you think that a `HashMap` is more appropriate in that case?

privat replied:
Oh. I did not thought about removal. You are right that in the Nit Array, a random removal is linear. However, if I look at `remove_vertex` I see that you do a loop `for i in [0..num_vertices[ do` anyway to update the matrix.
Maybe there is some way to use a freelist and just let a hole in the index_to_vertex...

privat commented on lib/graphs/digraph.nit:
```diff
@@ -0,0 +1,606 @@
+	end
+end
+
+private class QueuePair[V]
```
unused class?

R4PaSs commented on lib/graphs/digraph.nit:
```diff
@@ -0,0 +1,606 @@
+	end
+
+	# Returns a GraphViz string representing this digraph.
+	fun to_graphviz: String
```
Since this function outputs in dot format, I'd suggest renaming it to `to_dot`
This is quite a standard way of calling the operation in Nit too, as it is called this way in classes like `TreeNode` or `Text`

R4PaSs commented on lib/graphs/digraph.nit:
```diff
@@ -0,0 +1,606 @@
+# Implementations for representing directed graphs, also called digraphs.
+#
+# Currently, there are two classes
```
I know this is a WIP, but later on, could you add some precisions on the classes of the module or remove this comment since the classes document themselves ?

Morriar replied:
I think that `there are two classes` is not a really helpful information. Furthermore, it can be outdated if you add more classes and `nitdoc` already list those classes for you.

The interesting thing to mention here would be what class or classes should I look at first to use your library?

Morriar replied:
Also we have the habits to give a small usage example of the module.

R4PaSs commented on lib/graphs/digraph.nit:
```diff
@@ -0,0 +1,606 @@
+	# Returns true if and only if `u` is a predecessor of `v`.
+	fun is_predecessor(u, v: V): Bool
+	do
+		return has_arc(u, v)
```
Such small functions can be one-lined

R4PaSs commented on lib/graphs/digraph_test.nit:
```diff
@@ -0,0 +1,29 @@
+g.add_arc("d", "e")
+
+var path = g.a_shortest_path("a", "e")
+if path != null then print(path) else print "null"
```
I may be nitpicking, but as with the `print "null"`, I'd believe that the `print path` would not need parentheses, but I might be wrong, that'd need to be confirmed via testing

Morriar replied:
The main problem with prints is that it can be automatically checked, you should use asserts instead.

ablondin commented:
@privat and @R4PaSs: Thank you both for the review, this really helps!

Comments I agree with:

- [ ] Should be `AbstractDigraph[V]` instead of `AbstractDigraph[V: Object]`
- [ ] No newline in `to_s`
- [ ] Call `escape_to_dot` in `to_graphviz`
- [ ] Use a more efficient data structure for the queue in `a_shortest_path`
- [ ] `HashSet[V]` instead of `HashMap[V, Bool]`
- [ ] Change `loop-end` by `while-do-end` in `a_shortest_path` and `distance` methods
- [ ] `index_to_vertex: HashMap[Int, V]` should be `index_to_vertex: Array[V]`
- [ ] Remove useless `QueuePair` class
- [ ] Rename `to_graphviz` by `to_dot`
- [ ] One-line short functions

Comments I would like to discuss:

- Additional class for `preds` and `succs`. @privat: I see your point, but since these variables are private, I feel that an additional indirection is not necessary and would make the code less efficient for a small gain in readability. The concrete class `HashMapGraph` should not be touched much after defining the basic operations. I would tend to keep it as it is now.
- `fun vertices: RemovableCollection[V]` could make users invalidate the data structure. I agree that the user should not have the possibility to change the data structure. @privat: What would be the best solution? Clone it? Or maybe this is a case where a read-only iterator should be provided to iterate over each vertex.

privat commented:
Other idea: distinguish the read and the readwrite interface (as it is done with some collections for instance).
This way, most algorithms could be implemented in the read version (because they only need to query the graph) thus allowing ad hoc user-defined subclasses to only provide the query methods (I'm not sure if what I am saying is clear).

A use case could be some specific user data classes witch can be seen as a graph (eg a graph of wiki pages in a wiki engine). Then the user can chose to implement a view of this data structure (a view is a thin object that fulfill a specific interface but redirect the queries to real objects). While the view can redirect `has_vertex` or `has_arc` to the real data structure, it will be unpratical to have to implement `add_vertex` and cie because adding a vertex must pass trough the real data structure.

privat commented on lib/graphs/digraph.nit:
```diff
@@ -0,0 +1,606 @@
+	# g.add_arc(2, 3)
+	# assert g.num_arcs == 2
+	# ~~~
+	var num_arcs = 0
```
Maybe `num_vertices` and `num_arcs` should be abstract methods, and the subclasses should have to provide specific implementation (or just `redef var num_arc = 0` if they just want an attribute).
But I think it is better to let the choice of the implementation to subclasses.

For instance, in the 2 subclasses `num_vertices` could be implemented respectively by

~~~nit
redef fun num_vertices do return pred_map.length
redef fun num_vertices do return vertex_to_index.length
~~~

ablondin commented:
Updated task list:

- [ ] Should be `AbstractDigraph[V]` instead of `AbstractDigraph[V: Object]`
- [ ] No newline in `to_s`
- [ ] Call `escape_to_dot` in `to_graphviz`
- [ ] Use a more efficient data structure for the queue in `a_shortest_path`
- [ ] `HashSet[V]` instead of `HashMap[V, Bool]`
- [ ] Change `loop-end` by `while-do-end` in `a_shortest_path` and `distance` methods
- [ ] `index_to_vertex: HashMap[Int, V]` should be `index_to_vertex: Array[V]`
- [ ] Remove useless `QueuePair` class
- [ ] Rename `to_graphviz` by `to_dot`
- [ ] One-line short functions
- [ ] Change `vertices` method's signature (should return `Collection[V]`)
- [ ] `num_vertices` and `num_arcs` should be abstract methods
- [ ] Divide the services between *read-only* and *read/write*

ablondin commented:
Updated task list:

- [ ] `index_to_vertex: HashMap[Int, V]` should be `index_to_vertex: Array[V]`
- [ ] `num_vertices` and `num_arcs` should be abstract methods
- [ ] Divide the services between *read-only* and *read/write*

Morriar commented on lib/graphs/digraph.nit:
```diff
@@ -0,0 +1,561 @@
+module digraph
+
+# An arc of a digraph
+class Arc[V]
```
`V` alone with no bound means `nullable Object`. Does an arc can have only one end?

I will go for a bounded type on Object:

~~~nit
class Arc[V: Object]
~~~

Morriar commented on lib/graphs/digraph.nit:
```diff
@@ -0,0 +1,561 @@
+	# ~~~
+	var target: V
+
+	redef fun to_s: String do return "({source.to_s}, {target.to_s})"
```
Return type is not needed on not covariant methods.

If you chose to keep the nullable bound you should add a check before. Furthermore, I think that `to_s` is automatically applied on something that is not a string located in a superstring so you can change your implementation by:

~~~nit
redef fun to_s do return "({source}, {target})"
~~~

Morriar replied:
If you keep the nullable bound you can make your `to_s` safe with:

~~~nit
redef fun to_s do return "({source or else "NULL"}, {target or else "NULL"})"
~~~

Morriar commented on lib/graphs/digraph.nit:
```diff
@@ -0,0 +1,561 @@
+	redef fun to_s: String do return "({source.to_s}, {target.to_s})"
+end
+
+# Interface for digraphs
```
For the less mathsavvy of us, maybe you can add a little explanation of what is a digraphs and why it is usefull.

Morriar commented on lib/graphs/digraph.nit:
```diff
@@ -0,0 +1,561 @@
+
+	## ---------- ##
+	## Properties ##
+	## ---------- ##
```
I'm not sure of the relevance of this group. In Nit you generally want to group things by concerns.

Morriar commented on lib/graphs/digraph.nit:
```diff
@@ -0,0 +1,561 @@
+	var num_arcs = 0
+
+	## ---------------- ##
+	## Abstract methods ##
```
e

Morriar commented on lib/graphs/digraph.nit:
```diff
@@ -0,0 +1,561 @@
+	fun successors(u: V): Collection[V] is abstract
+
+	# Returns the vertices of this graph.
+	fun vertices: Collection[V] is abstract
```
Maybe you can group that methods and `arcs` with `num_arc` and `num_vertices` under the name `Graph data`

Morriar commented on lib/graphs/digraph.nit:
```diff
@@ -0,0 +1,561 @@
+
+	## ------------ ##
+	## Neighborhood ##
+	## ------------ ##
```
This group is useful :)

Morriar commented on lib/graphs/digraph.nit:
```diff
@@ -0,0 +1,561 @@
+# Implementations for representing directed graphs, also called digraphs.
+#
```
Maybe you can add some information about the digraph concept here.

An interesting example can be found here: http://networkx.lanl.gov/reference/classes.digraph.html

Morriar commented on lib/graphs/digraph.nit:
```diff
@@ -0,0 +1,561 @@
+class ArrayDigraph[V: Object]
+	super AbstractDigraph[V]
+
+	# Attributes
```
Is you don't add a block line after this comment it will be attached to `vertex_to_index`

privat commented on lib/graphs/digraph.nit:
```diff
@@ -0,0 +1,561 @@
+	end
+
+	# ------- #
+	# Degrees #
```
Should degree be something related to Neighborhood?

ablondin commented:
Pull request is now almost finished.

I took into account all comments below. I also changed arcs so that they now store values by default (`Arc[V]` is now `Arc[V,A]`). I documented each function.

To make the contribution smaller, I also removed one of the concrete implementation, which was not working anymore. I plan to add it later.

The only elements remaining now are the following:
[ ] Document the two modules
[ ] Remove the `digraph_test.nit` file

I'm also dissatisfied with the `add_arcs` function:

~~~
# ~~~                                                        
# import digraph                                             
# var g = new HashMapDigraph[Int, nullable Int]              
# var arcs = new Array[Arc[Int, nullable Int]]               
# arcs.add(new Arc[Int, nullable Int](0, 1))                 
# arcs.add(new Arc[Int, nullable Int](1, 2))                 
# arcs.add(new Arc[Int, nullable Int](1, 2))                 
# g.add_arcs(arcs)                                           
# assert g.num_arcs == 2                                     
# ~~~                                                        
fun add_arcs(arcs: Collection[Arc[V, A]])                    
do                                                           
        for a in arcs do add_arc(a.source, a.target, a.value)
end                                                          
~~~

Do you have any idea how one could add multiple arcs in one line, something like:
~~~
import digraph                                             
var g = new HashMapDigraph[Int, nullable Int]              
g.add_arcs([[0,1],[1,2],[2,3]])
~~~
seems a good idea, but I would have to check if each small lists has 2 or 3 elements according to whether there is a value on the arc or not.

privat commented on lib/graphs/digraph.nit:
```diff
@@ -0,0 +1,147 @@
+		if has_vertex(u) and has_vertex(v) then
+			for arc in outgoing_arcs_map[u] do
+				if arc.target == v then return arc.value
+			end
```
I suggest that you add a private `intern_get_arc` that returns the `nullable Arc` that you can use to factorize some code in `has_arc`, `get_arc_value` or `remove_arc`

privat commented on lib/graphs/abstract_digraph.nit:
```diff
@@ -0,0 +1,647 @@
+	# assert g.has_arc(1, 2)
+	# assert not g.has_arc(1, 0)
+	# ~~~
+	fun add_arc(u, v: V, l: nullable A) is abstract
```
I understand that adding an existing vertex is a no-op but the behavior is unclear for arcs.Zhat should do the following?

~~~nit
var g = new HashMapDigraph[Int, nullable String]
g.add_arc(0, 1, "Hello")
g.add_arc(0, 1, "World")
print g.get_arc_value(0, 1) # "Hello" or "World"
~~~

privat commented on lib/graphs/abstract_digraph.nit:
```diff
@@ -0,0 +1,647 @@
+	# g.change_arc_value(0, 1, 4)
+	# assert g.get_arc_value(0, 1) == 4
+	# ~~~
+	fun change_arc_value(u, v: V, l: A) is abstract
```
Hum, what about Nit's generalized setters?

They could allow the user to write:

~~~nit
graph.arc_value(x, y) = 10
print graph.arc_value(x, y) # => 10
~~~

Just rename `get_arc_value` to `arc_value` (in nit, getters do not require the `get` prefix) and rename `change_arc_value` to `arc_value=` (the `=` belong to the name). eg:

~~~
class G
   fun arc_value(u, v: V): A is abstract
   fun arc_value=(u, v: V, l: A) is abstract
end
~~~

The advantage of generalized setter is that it simplify the vocabulary (only one identifier to remember). Another advantage is magic combined assignment, free of charge.

~~~nit
graph.arc_value(x, y) += 1
# equivalent to
graph.arc_value(x, y) = graph.arc_value(x, y) + 1
~~~

privat commented on lib/graphs/digraph.nit:
```diff
@@ -0,0 +1,147 @@
+		end
+	end
+
+	redef fun outgoing_arcs(u: V): nullable Collection[Arc[V, A]]
```
in `arcs`, `incoming_arcs` and `outgoing_arcs` the Arc objects used internally escape, this might cause the graph to become inconsistent if they are used badly.

Some solutions (not exhaustive, and contain bad idea):

* document the fact that the returned arcs are ones used internally, and that with great power comes great responsibility
* returns copies of the arc objects
* add a read interface for arcs (ReadArc?) and use it instead in the signature

ablondin replied:
Ideally, we should avoid the first solution.

I don't really like the second idea, which would create new objects for bad reason.

So that leaves me with the third solution. Isn't there a way to just "cast" the collection to something like ReadCollection without having to define it?

ablondin commented:
@privat : Thank you for the good comments. I'll update a fix soon.

Morriar commented:
Maybe the Digraph abstraction can implement some kind of collection.

I'm thinking at some useful services like `is_empty` or `iterator`.

privat commented on lib/graphs/abstract_digraph.nit:
```diff
@@ -0,0 +1,583 @@
+module abstract_digraph
+
+# Interface for digraphs
+abstract class AbstractDigraph[V]
```
I suggest to remove the `Abstract` part of the name as it conveys more noise that information.

ablondin replied:
I agree. What about naming it ``Digraph`` and the concrete implementation ``HashMapDigraph``?

privat replied:
Yes. `HashMapDigraph` or maybe just `HashDigraph` as it rely on a working and efficient implementation of `hash` in vertices. Map is important but maybe not that important to be in the name (it should be in the doc)

privat commented on lib/graphs/digraph.nit:
```diff
@@ -0,0 +1,114 @@
+		return outgoing_vertices_map.keys
+	end
+
+	redef fun arcs: Collection[Array[V]]
```
I think that the implementation of `arcs`, `incoming_arcs` and `outgoing_arcs` can be moved up as a sane default implementation in the abstract class

privat commented:
nice. I do not see more things to say.

ablondin commented:
@privat: A last question. I decided to bound the generic type ``V`` by ``Object``, since allowing nullable objects would make some algorithms more complicated for almost no gain. Does it seem reasonable to you?

privat commented:
`Object` instead of `nullable Object` is fine enough

privat commented:
A lot of nitunits are failing. You should consider checking them with the `nitunit` tool.

ablondin commented:
@privat: I know. But I can't use ``nitunit`` right now since it's broken on MacOS (same bug as in issue #1623).

ablondin commented:
Sorry, I was wrong, it actually works, this was a different problem. I mean, there is still the bug of issue #1623, but it doesn't prevent me from using ``nitunit``.

ablondin commented:
@privat: I've got an error that I don't know how to fix. Now, all unit tests pass except for 3 due to the same error:
~~~
abstract_digraph.nit:141,2--153,0: ERROR: nitunit.abstract_digraph.abstract_digraph::AbstractDigraph.abstract_digraph::AbstractDigraph::arcs_itera
tor (in .nitunit/abstract_digraph-10.nit): Runtime error: Uninitialized attribute _graph (abstract_digraph.nit:527)
-------------------------------------------------
--   Stack Trace   ------------------------------
-------------------------------------------------
` fatal_exit
` abstract_digraph#ArcsIterator#graph (abstract_digraph.nit:527)
` NEW_abstract_digraph__ArcsIterator
` abstract_digraph#AbstractDigraph#arcs_iterator (abstract_digraph.nit:141)
` abstract_digraph-10#standard::Sys#main (.nitunit/abstract_digraph-10.nit:6)
` main
-------------------------------------------------
 (error)
                # Returns an iterator over the arcs of this graph
                ^
~~~

As I understand it, it says that the ``graph`` attribute of ``ArcsIterator`` is not initialized when I call it from the method ``arcs_iterator`` of ``AbstractDigraph``. However, it is not the case, since I call it with ``self`` as an argument for ``graph``. What am I missing here?

privat commented:
@ablondin this is a common error about object initialization when the initial value of an attribute rely on the value of another attribute. The order of the initialization is therefore important.

The current specification about object construction is that attributes with a default value are initialized before attributes that are user-initialized.
In your code, `ArcsIterator::graph` is user-initialized, but `ArcsIterator::sources_iterator` have a default value. therefore `sources_iterator` is initialized first.
Unfortunately it requires `graph` that is not yet initialized.

For more details, the initialization of attributes with a default value is done in a C function called `NEW_$classname` that is the first step of object construction.
You can see in the stack trace that `graph` is invoked by `NEW_abstract_digraph__ArcsIterator`. It means that an attribute with a default value called `graph` before the first step of initialization is finished. It is then easy to look in the definition of the class if there is some attributes that use `graph` in their default value.

There is several way to solve this problem. A simple one is to remove the default value from `sources_iterator` and manually initializes it in the `init`; so something like:

~~~nit
private var sources_iterator: Iterator[V] is noinit
init do
   sources_iterator = graph.vertices_iterator
end
~~~

Another solution is to use the unpopular annotation `lateinit`.

~~~
private var sources_iterator: Iterator[V] = graph.vertices_iterator is lateinit
~~~

This annotation explains that the default value has to be computed in the second step of the initialization, with user-initialized attributes.

The first reason that this annotation is unpopular is that the order of attributes in the class is important: `sources_iterator` must be declared after `graph` (or else you will have the same error).
The second reason that this annotation is unpopular is that it might cause un-POLA side effects in subclasses that tries to use the `autoinit` annotation.

ablondin commented:
A small update:

* Iterators working
* Classes renamed
* Minor restructuration

The only remaining point is adding module documentation!

privat commented:
great!

ablondin commented:
@Morriar: I started documenting the module. It is not complete yet, but I have some questions before I can continue.

1. How can I reference a picture to be included in the documentation. I tried it (check in the header of the ``digraph.nit`` file, where I follow the Markdown syntax), but it doesn't work.
2. How can I add a link to some class. For instance, if I want to tell the reader to look at class X for further details, how can I create a real link to the documentation of this class?
3. I first documented the module in a README file, but I don't know where to find the result of this particular file when I generate the documentation with ``nitdoc``. Where should I look?
4. How much should I document the class? I can write a small tutorial on graph theory, but maybe it's too much...

Morriar commented:
@ablondin:

1. Images are recognized by markdown like you used them, but for now we don't import external files into the outputed documentation. An easy workaround would be to link directly the image to Github.

2. There is three possible ways to do that:
    * `[[MyClass]]`
    * ``` `MyClass` ```
    * `MyClass`

Feel free to commit a beta version of the documentation if you want us to give an eye.
    The three do the same thing but only the first one is implemented in READMEs right now :)

3. READMEs are used as the default documentation for a group.
    For example, the standard/README.md is rendered as http://nitlanguage.org/doc/stdlib/group_standard.html
    If you want to access the generated API documentation (not the one from the readme) you should check at http://nitlanguage.org/doc/stdlib/api_group_standard.html

4. When documenting classes, the most usefull informations are:
    * What does this class?
    * How can I instantiate this class (example)? 
    * What are the main methods to look at.

   In your case, I guess that the small tutorial on graph theory can be written in the module documentation since it's the one that introduce all the concepts in the module.

ablondin commented:
Thanks @Morriar. Could you elaborate more about linking images on Github? I'm not sure I understand what you mean? Using an absolute link? I think the links on Github change with the version, so that I wouldn't be safe...

Morriar commented:
@ablondin You can use the link to master one the project is merged it should be stable enough untill we know what to do with external resources files.

~~~
https://github.com/privat/nit/blob/master/my/file.png
~~~

ablondin commented:
Just pushed another version, which should be final. Here are the latest changes:

- I moved the part computing the strongly connected components with Tarjan's method to a private class, to avoid creating private attributes for any digraph, even if it doesn't use the method;
- I squashed all commits to a single one;
- I finished documenting the module (@Morriar: could you check it again if you have time);
- I added a license;
- I removed useless whitespaces.

Morriar commented on lib/graphs/digraph.nit:
```diff
@@ -0,0 +1,846 @@
+# It is therefore simple to provide another implementation: One only has to
+# create a nonabstract specialization of either `Digraph` or `MutableDigraph`.
+#
+# Basic usage
```
I was expecting a section `Advanced Usage` after this one :)

ablondin replied:
I should add a comment saying that for more advanced usage, one should look at the docstrings.

Morriar commented on lib/graphs/digraph.nit:
```diff
@@ -0,0 +1,846 @@
+# Then we can add vertices and arcs. Note that if an arc is added whose source
+# and target are not already in the digraph, the vertices are added beforehand.
+# ~~~
+# import digraph
```
Niitdoc is supposed to be smart enought to concatenate code blocks if it can do it.

Therefore, `g` stills declared here and you can reuse it in your example. If you put the `import` again, Nitdoc will detect that this cannot be concatenated with the previous example.

It all depends on what you want to do / show with your example. But know that you can save a few lines with this feature and provide more precise examples.

Morriar replied:
Also, I don't think it need to be changed.

ablondin replied:
When you say that the `import`s don't need to be repeated, it's not true for the unit tests, right? Only the code blocks in the same block of documentation (like the module header)?

Morriar replied:
No, it's also true with unit tests (it's the same process).

An example here: http://nitlanguage.org/doc/stdlib/property_standard__collection__array__Array___43d.html#standard__collection__array__Array___43d

In the given example, a1, a2, a3 are defined by the first block and used in assert of the second blocks. Nitdoc/Nitunit merges the two blocks and run them as one.

Morriar commented on lib/graphs/digraph.nit:
```diff
@@ -0,0 +1,846 @@
+# # Then call "dot -Tpng -o graph.png"
+# ~~~
+#
+# ![A graph drawing produced by Graphviz](https://github.com/privat/nit/blob/master/lib/graph.png)
```
The module documentation `as it` already meet the Nit quality criteria and can be merged.

But there is also place for improvment. It all depends how far in the documentation you are willing to go :)

Some questions that should be anwsered at this point (because when I'm looking for the perfect module for my needs, I don't want to read all the code):

* can I associate arbitrary values to nodes/arcs?
* is there high level services (time to speack about cycles, shortest paths etc...
* how can I implement new features? subclassing what and how? what virtual types are involved in such case?

Also, if there still some points on your todo list that you don't whant to implement before you pull request, you can save them here in a section `# TODO`.

Morriar commented:
Pretty and cool, +1

ablondin commented:
There are two failed checks and I don't understand them...

- "Signed-off", which is in the commit?!?
- The license added in the header seems to make `nitunit` fail. Why is it wrong? I copied/pasted the header of another file, so it should be ok, right?

Morriar commented:
For the header, remove the dash at Line 16. Sorry Imiss this in the review.

privat commented:
for the signed-off-by issue: the author is

* Signed-off-by: Alexandre Blondin Massé <alexandre.blondin.masse@gmail.com>

but the signed-off-by line is

* Signed-off-by: Alexandre Blondin Massé <blondin_masse.alexandre@uqam.ca>

Since the the email is not the same, jenkins complains

privat commented:
+1 for me

ablondin commented:
Reroll:

- @Morriar: I took your comments about documentation intro consideration. Please tell me if it seems ok to you;
- License check should be ok now;
- Signed-off-by should be ok also.

Morriar commented on lib/graphs/digraph.nit:
```diff
@@ -0,0 +1,907 @@
+# Extending the library
+# =====================
+#
+# There at least two ways of providing new methods on digraphs. If the method
```
There are* ?

privat commented:
Tests are fine, I was playing with #1639 that has a side effect on the last run

ablondin commented:
Reroll: 

- Last typo suggested by Morriar
- Removed useless png file from directory.

ablondin commented:
Finally, all tests pass!
